utils/http/proxy/proxy.go

```diff
@@ -0,0 +1,153 @@
+package proxy
+
+import (
+	"bytes"
+	"context"
+	"io"
+	"net/http"
+	"strconv"
+	"strings"
+
+	"github.com/go-http-utils/headers"
+
+	"github.com/ARM-software/golang-utils/utils/commonerrors"
+	httpheaders "github.com/ARM-software/golang-utils/utils/http/headers"
+	"github.com/ARM-software/golang-utils/utils/reflection"
+	"github.com/ARM-software/golang-utils/utils/safecast"
+	"github.com/ARM-software/golang-utils/utils/safeio"
+)
+
+// ProxyDisallowList describes headers which are not proxied back.
+var ProxyDisallowList = []string{
+	headers.AccessControlAllowOrigin,
+	headers.AccessControlAllowMethods,
+	headers.AccessControlAllowHeaders,
+	headers.AccessControlExposeHeaders,
+	headers.AccessControlMaxAge,
+	headers.AccessControlAllowCredentials,
+}
+
+// ProxyRequest proxies a request to a new endpoint. The method can also be changed. Headers are sanitised during the process.
+func ProxyRequest(r *http.Request, proxyMethod, endpoint string) (proxiedRequest *http.Request, err error) {
+	if reflection.IsEmpty(r) {
+		err = commonerrors.UndefinedVariable("request to proxy")
+		return
+	}
+	ctx := r.Context()
+	contentLength := determineRequestContentLength(r)
+	h := httpheaders.FromRequest(r).AllowList(headers.Authorization)
+	if reflection.IsEmpty(proxyMethod) {
+		proxyMethod = http.MethodGet
+	}
+	proxiedRequest, err = http.NewRequestWithContext(ctx, proxyMethod, endpoint, r.Body)
+	if err != nil {
+		err = commonerrors.WrapError(commonerrors.ErrUnexpected, err, "could not create a proxied request")
+		return
+	}
+
+	if proxiedRequest.ContentLength <= 0 {
+		if proxiedRequest.Body == nil || proxiedRequest.Body == http.NoBody {
+			if contentLength > 0 {
+				proxiedRequest.Body = r.Body
+				proxiedRequest.GetBody = r.GetBody
+			} else {
+				proxiedRequest, err = http.NewRequestWithContext(ctx, proxyMethod, endpoint, convertBody(ctx, r.Body))
+				if err != nil {
+					err = commonerrors.WrapError(commonerrors.ErrUnexpected, err, "could not create a proxied request")
+					return
+				}
```

Comment:
Behaviourally I don't really understand what is going on here (could be lack of experience). Why will it perform a re-request, also why is the proxied request body replaced with the original request body?

Reply:
Here, I am trying to consider all the weird edge cases which can happen because of go implementation and also trying to reduce the amount of copying so that there is no penalty in performance